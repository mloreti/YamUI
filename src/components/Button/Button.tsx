/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { Button as FabricButton } from 'office-ui-fabric-react/lib/components/Button/Button';
import { BaseButton } from 'office-ui-fabric-react/lib/components/Button/BaseButton';
import { BaseComponentProps } from '../../util/BaseComponent/props';
import Block, { TextSize } from '../Block';
import Spinner, { SpinnerColor, SpinnerSize } from '../Spinner';
import { ButtonColor, ButtonStatus, ButtonIconPosition, ButtonSize, ButtonType } from './enums';
import BaseIcon from '../Icon/BaseIcon';
import './Button.css';

export { ButtonColor, ButtonStatus, ButtonIconPosition, ButtonSize, ButtonType };

const hrefBlacklist = ['', '#', 'javascript://'];
export type FabricButtonType = HTMLAnchorElement | HTMLDivElement | HTMLButtonElement | BaseButton | FabricButton;
export interface BaseButtonProps extends BaseComponentProps {
  /**
   * Visible button text.
   */
  text: string;

  /**
   * Additional label that must be provided if the button text is not descriptive enough.
   */
  ariaLabel?: string;

  /**
   * Stretch the button to fill the available horizontal space.
   */
  fullWidth?: boolean;

  /**
   * Color describing the button's intent.
   * @default ButtonColor.SECONDARY
   */
  color?: ButtonColor;

  /**
   * Optional icon.
   */
  icon?: typeof BaseIcon;

  /**
   * Icon position.
   * @default ButtonIconPosition.LEFT
   */
  iconPosition?: ButtonIconPosition;

  /**
   * Button size.
   * @default ButtonSize.REGULAR
   */
  size?: ButtonSize;

  /**
   * Button type.
   * @default ButtonType.BUTTON
   */
  type?: ButtonType;

  /**
   * Click callback handler.
   */
  onClick?: React.MouseEventHandler<FabricButtonType>;

  /**
   * Focus callback handler.
   */
  onFocus?: React.FocusEventHandler<FabricButtonType>;

  /**
   * Blur callback handler.
   */
  onBlur?: React.FocusEventHandler<FabricButtonType>;

  /**
   * Mouse enter callback handler.
   */
  onMouseEnter?: React.MouseEventHandler<FabricButtonType>;

  /**
   * Mouse leave callback handler.
   */
  onMouseLeave?: React.MouseEventHandler<FabricButtonType>;
}

export interface RegularButtonProps extends BaseButtonProps {
  /**
   * Buttons cannot have a link.
   */
  href?: void;

  /**
   * Status of this button.
   * @default ButtonStatus.ENABLED
   */
  status?: ButtonStatus;

  /**
   * Screenreader text for loading state.
   */
  loadingText?: string;
}

export interface LoadingButtonProps extends RegularButtonProps {
  status: ButtonStatus.LOADING;
  loadingText: string;
}

export interface LinkButtonProps extends BaseButtonProps {
  /**
   * URL or a URL fragment that the link points to. If provided, the component will be rendered as a
   * link styled as a button.
   */
  href: string;

  /**
   * Links cannot be disabled nor loading.
   */
  status?: void;

  /**
   * Links cannot have a type.
   */
  type?: undefined;
}

export type ButtonProps = RegularButtonProps | LoadingButtonProps | LinkButtonProps;

/**
 * A `Button` allows a user to take an action.
 */
export default class Button extends React.Component<ButtonProps> {
  public static propTypes = {
    // TypeScript does not support negated types, so we need to do a runtime validation instead.
    href(props: LinkButtonProps, propName: string, componentName: string) {
      const href = props.href;
      if (typeof href !== 'string') {
        return;
      }
      if (hrefBlacklist.indexOf(href) > -1) {
        return new Error(
          `Invalid prop ${propName} supplied to ${componentName}: "${href}".
          Please use a Button for actions, NavigationLink (or Button with href) for navigation.`,
        );
      }
    },
  };

  public static defaultProps = {
    color: ButtonColor.SECONDARY,
    iconPosition: ButtonIconPosition.LEFT,
    size: ButtonSize.REGULAR,
    status: ButtonStatus.ENABLED,
    type: ButtonType.BUTTON,
  };

  public render() {
    const { ariaLabel, type, onClick, onFocus, onBlur, onMouseEnter, onMouseLeave } = this.props;

    const href = (this.props as LinkButtonProps).href;
    const status = (this.props as RegularButtonProps).status;
    const isDisabled = status === ButtonStatus.DISABLED || status === ButtonStatus.LOADING;

    return (
      <BaseButton
        ariaLabel={ariaLabel}
        className={this.getClasses()}
        disabled={isDisabled}
        href={href}
        type={href ? '' : type}
        onBlur={onBlur}
        onClick={onClick}
        onFocus={onFocus}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
      >
        {this.getContents()}
        {this.isLoading() && this.getSpinner()}
      </BaseButton>
    );
  }

  private getContents() {
    const { icon: Icon, iconPosition, size, text } = this.props;

    const textSize = size === ButtonSize.SMALL ? TextSize.SMALL : TextSize.MEDIUM_SUB;

    const buttonIcon = Icon && (
      <span className={`y-button--icon-wrapper-${iconPosition}`}>
        <Icon />
      </span>
    );

    const className = this.isLoading() ? 'y-button--content__hidden' : '';

    return (
      <Block textSize={textSize} className={className}>
        {iconPosition === 'left' && buttonIcon}
        {text}
        {iconPosition === 'right' && buttonIcon}
      </Block>
    );
  }

  private getSpinner() {
    const { color, size } = this.props;

    const loadingText = (this.props as LoadingButtonProps).loadingText;

    const spinnerColor = color === ButtonColor.PRIMARY ? SpinnerColor.DARK : SpinnerColor.LIGHT;
    const spinnerSize = size === ButtonSize.SMALL ? SpinnerSize.XSMALL : SpinnerSize.SMALL;

    return (
      <span className="y-button--spinner">
        <Spinner color={spinnerColor} size={spinnerSize} text={loadingText} hideText={true} isCentered={true} />
      </span>
    );
  }

  private getClasses() {
    const { className, color, status, size, fullWidth } = this.props;

    const classes: string[] = ['y-button', `y-button__color-${color}`, `y-button__size-${size}`];
    if (status !== ButtonStatus.ENABLED) {
      classes.push(`y-button__state-${status}`);
    }
    if (fullWidth) {
      classes.push('y-button__fullWidth');
    }
    if (className) {
      classes.push(className);
    }

    return classes.join(' ');
  }

  private isLoading() {
    return this.props.status === ButtonStatus.LOADING;
  }
}
