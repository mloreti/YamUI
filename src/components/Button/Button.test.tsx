/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import Button, {
  ButtonProps,
  ButtonColor,
  ButtonStatus,
  ButtonIconPosition,
  ButtonSize,
  ButtonType,
} from '.';
import AccountsIcon from '../Icon/icons/Accounts';

const sampleText = 'Click Me';
const sampleAriaLabel = 'Aria description';

describe('<Button />', () => {
  let component: ShallowWrapper<ButtonProps>;

  describe('with default options', () => {
    beforeEach(() => {
      component = shallow(<Button text={sampleText} />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with button type', () => {
    beforeEach(() => {
      component = shallow(<Button text={sampleText} type={ButtonType.BUTTON} />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with submit type', () => {
    beforeEach(() => {
      component = shallow(<Button text={sampleText} type={ButtonType.SUBMIT} />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with reset type', () => {
    beforeEach(() => {
      component = shallow(<Button text={sampleText} type={ButtonType.RESET} />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with additional className', () => {
    beforeEach(() => {
      component = shallow(<Button text={sampleText} className="TEST_CLASSNAME" />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with aria label', () => {
    beforeEach(() => {
      component = shallow(<Button text={sampleText} ariaLabel={sampleAriaLabel} />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with fullWidth', () => {
    beforeEach(() => {
      component = shallow(<Button text={sampleText} fullWidth={true} />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('set as disabled', () => {
    beforeEach(() => {
      component = shallow(<Button text={sampleText} status={ButtonStatus.DISABLED} />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('set as loading', () => {
    beforeEach(() => {
      component = shallow(<Button text={sampleText} status={ButtonStatus.LOADING} />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with icon', () => {
    beforeEach(() => {
      component = shallow(<Button icon={AccountsIcon} text={sampleText} />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });

    describe('on the right', () => {
      beforeEach(() => {
        component = shallow(<Button icon={AccountsIcon} text={sampleText} iconPosition={ButtonIconPosition.RIGHT} />);
      });

      it('matches its snapshot', () => {
        expect(component).toMatchSnapshot();
      });
    });
  });

  describe('with primary color', () => {
    beforeEach(() => {
      component = shallow(<Button color={ButtonColor.PRIMARY} text={sampleText} />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });

    describe('set as loading', () => {
      beforeEach(() => {
        component = shallow(<Button color={ButtonColor.PRIMARY} text={sampleText} status={ButtonStatus.LOADING} />);
      });

      it('matches its snapshot', () => {
        expect(component).toMatchSnapshot();
      });
    });
  });

  describe('with small size', () => {
    beforeEach(() => {
      component = shallow(<Button size={ButtonSize.SMALL} text={sampleText} />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });

    describe('with icon', () => {
      beforeEach(() => {
        component = shallow(<Button icon={AccountsIcon} size={ButtonSize.SMALL} text={sampleText} />);
      });

      it('matches its snapshot', () => {
        expect(component).toMatchSnapshot();
      });
    });

    describe('set as loading', () => {
      beforeEach(() => {
        component = shallow(<Button size={ButtonSize.SMALL} text={sampleText} status={ButtonStatus.LOADING} />);
      });

      it('matches its snapshot', () => {
        expect(component).toMatchSnapshot();
      });
    });
  });

  describe('with valid href', () => {
    beforeEach(() => {
      component = shallow(<Button text={sampleText} href="https://www.yammer.com" />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with invalid href', () => {
    beforeEach(() => {
      component = shallow(<Button text={sampleText} href="#" />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with event callbacks', () => {
    let onClick: jest.Mock;
    let onMouseEnter: jest.Mock;
    let onMouseLeave: jest.Mock;
    let onFocus: jest.Mock;
    let onBlur: jest.Mock;

    beforeEach(() => {
      onClick = jest.fn();
      onMouseEnter = jest.fn();
      onMouseLeave = jest.fn();
      onFocus = jest.fn();
      onBlur = jest.fn();
      component = shallow(
        <Button
          text={sampleText}
          onClick={onClick}
          onMouseEnter={onMouseEnter}
          onMouseLeave={onMouseLeave}
          onFocus={onFocus}
          onBlur={onBlur}
        />,
      );
    });

    describe('when clicked', () => {
      beforeEach(() => {
        component.simulate('click');
      });

      it('triggers its onClick callback', () => {
        expect(onClick).toHaveBeenCalled();
      });
    });

    describe('when hovered', () => {
      beforeEach(() => {
        component.simulate('mouseEnter');
      });

      it('triggers its onMouseEnter callback', () => {
        expect(onMouseEnter).toHaveBeenCalled();
      });
    });

    describe('when hovered out', () => {
      beforeEach(() => {
        component.simulate('mouseLeave');
      });

      it('triggers its onMouseLeave callback', () => {
        expect(onMouseLeave).toHaveBeenCalled();
      });
    });

    describe('when focused', () => {
      beforeEach(() => {
        component.simulate('focus');
      });

      it('triggers its onFocus callback', () => {
        expect(onFocus).toHaveBeenCalled();
      });
    });

    describe('when blurred', () => {
      beforeEach(() => {
        component.simulate('blur');
      });

      it('triggers its onBlur callback', () => {
        expect(onBlur).toHaveBeenCalled();
      });
    });
  });
});
