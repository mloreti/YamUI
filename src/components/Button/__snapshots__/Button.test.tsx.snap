// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> set as disabled matches its snapshot 1`] = `
<BaseButton
  baseClassName="ms-Button"
  className="y-button y-button__color-secondary y-button__size-regular y-button__state-disabled"
  classNames={Object {}}
  disabled={true}
  split={false}
  styles={Object {}}
  type="button"
>
  <Block
    className=""
    textSize="mediumSub"
  >
    Click Me
  </Block>
</BaseButton>
`;

exports[`<Button /> set as loading matches its snapshot 1`] = `
<BaseButton
  baseClassName="ms-Button"
  className="y-button y-button__color-secondary y-button__size-regular y-button__state-loading"
  classNames={Object {}}
  disabled={true}
  split={false}
  styles={Object {}}
  type="button"
>
  <Block
    className="y-button--content__hidden"
    textSize="mediumSub"
  >
    Click Me
  </Block>
  <span
    className="y-button--spinner"
  >
    <Spinner
      color="light"
      hideText={true}
      isCentered={true}
      size="small"
    />
  </span>
</BaseButton>
`;

exports[`<Button /> with additional className matches its snapshot 1`] = `
<BaseButton
  baseClassName="ms-Button"
  className="y-button y-button__color-secondary y-button__size-regular TEST_CLASSNAME"
  classNames={Object {}}
  disabled={false}
  split={false}
  styles={Object {}}
  type="button"
>
  <Block
    className=""
    textSize="mediumSub"
  >
    Click Me
  </Block>
</BaseButton>
`;

exports[`<Button /> with aria label matches its snapshot 1`] = `
<BaseButton
  ariaLabel="Aria description"
  baseClassName="ms-Button"
  className="y-button y-button__color-secondary y-button__size-regular"
  classNames={Object {}}
  disabled={false}
  split={false}
  styles={Object {}}
  type="button"
>
  <Block
    className=""
    textSize="mediumSub"
  >
    Click Me
  </Block>
</BaseButton>
`;

exports[`<Button /> with button type matches its snapshot 1`] = `
<BaseButton
  baseClassName="ms-Button"
  className="y-button y-button__color-secondary y-button__size-regular"
  classNames={Object {}}
  disabled={false}
  split={false}
  styles={Object {}}
  type="button"
>
  <Block
    className=""
    textSize="mediumSub"
  >
    Click Me
  </Block>
</BaseButton>
`;

exports[`<Button /> with default options matches its snapshot 1`] = `
<BaseButton
  baseClassName="ms-Button"
  className="y-button y-button__color-secondary y-button__size-regular"
  classNames={Object {}}
  disabled={false}
  split={false}
  styles={Object {}}
  type="button"
>
  <Block
    className=""
    textSize="mediumSub"
  >
    Click Me
  </Block>
</BaseButton>
`;

exports[`<Button /> with fullWidth matches its snapshot 1`] = `
<BaseButton
  baseClassName="ms-Button"
  className="y-button y-button__color-secondary y-button__size-regular y-button__fullWidth"
  classNames={Object {}}
  disabled={false}
  split={false}
  styles={Object {}}
  type="button"
>
  <Block
    className=""
    textSize="mediumSub"
  >
    Click Me
  </Block>
</BaseButton>
`;

exports[`<Button /> with icon matches its snapshot 1`] = `
<BaseButton
  baseClassName="ms-Button"
  className="y-button y-button__color-secondary y-button__size-regular"
  classNames={Object {}}
  disabled={false}
  split={false}
  styles={Object {}}
  type="button"
>
  <Block
    className=""
    textSize="mediumSub"
  >
    <span
      className="y-button--icon-wrapper-left"
    >
      <Accounts />
    </span>
    Click Me
  </Block>
</BaseButton>
`;

exports[`<Button /> with icon on the right matches its snapshot 1`] = `
<BaseButton
  baseClassName="ms-Button"
  className="y-button y-button__color-secondary y-button__size-regular"
  classNames={Object {}}
  disabled={false}
  split={false}
  styles={Object {}}
  type="button"
>
  <Block
    className=""
    textSize="mediumSub"
  >
    Click Me
    <span
      className="y-button--icon-wrapper-right"
    >
      <Accounts />
    </span>
  </Block>
</BaseButton>
`;

exports[`<Button /> with invalid href matches its snapshot 1`] = `
<BaseButton
  baseClassName="ms-Button"
  className="y-button y-button__color-secondary y-button__size-regular"
  classNames={Object {}}
  disabled={false}
  href="#"
  split={false}
  styles={Object {}}
  type=""
>
  <Block
    className=""
    textSize="mediumSub"
  >
    Click Me
  </Block>
</BaseButton>
`;

exports[`<Button /> with primary color matches its snapshot 1`] = `
<BaseButton
  baseClassName="ms-Button"
  className="y-button y-button__color-primary y-button__size-regular"
  classNames={Object {}}
  disabled={false}
  split={false}
  styles={Object {}}
  type="button"
>
  <Block
    className=""
    textSize="mediumSub"
  >
    Click Me
  </Block>
</BaseButton>
`;

exports[`<Button /> with primary color set as loading matches its snapshot 1`] = `
<BaseButton
  baseClassName="ms-Button"
  className="y-button y-button__color-primary y-button__size-regular y-button__state-loading"
  classNames={Object {}}
  disabled={true}
  split={false}
  styles={Object {}}
  type="button"
>
  <Block
    className="y-button--content__hidden"
    textSize="mediumSub"
  >
    Click Me
  </Block>
  <span
    className="y-button--spinner"
  >
    <Spinner
      color="dark"
      hideText={true}
      isCentered={true}
      size="small"
    />
  </span>
</BaseButton>
`;

exports[`<Button /> with reset type matches its snapshot 1`] = `
<BaseButton
  baseClassName="ms-Button"
  className="y-button y-button__color-secondary y-button__size-regular"
  classNames={Object {}}
  disabled={false}
  split={false}
  styles={Object {}}
  type="reset"
>
  <Block
    className=""
    textSize="mediumSub"
  >
    Click Me
  </Block>
</BaseButton>
`;

exports[`<Button /> with small size matches its snapshot 1`] = `
<BaseButton
  baseClassName="ms-Button"
  className="y-button y-button__color-secondary y-button__size-small"
  classNames={Object {}}
  disabled={false}
  split={false}
  styles={Object {}}
  type="button"
>
  <Block
    className=""
    textSize="small"
  >
    Click Me
  </Block>
</BaseButton>
`;

exports[`<Button /> with small size set as loading matches its snapshot 1`] = `
<BaseButton
  baseClassName="ms-Button"
  className="y-button y-button__color-secondary y-button__size-small y-button__state-loading"
  classNames={Object {}}
  disabled={true}
  split={false}
  styles={Object {}}
  type="button"
>
  <Block
    className="y-button--content__hidden"
    textSize="small"
  >
    Click Me
  </Block>
  <span
    className="y-button--spinner"
  >
    <Spinner
      color="light"
      hideText={true}
      isCentered={true}
      size="xSmall"
    />
  </span>
</BaseButton>
`;

exports[`<Button /> with small size with icon matches its snapshot 1`] = `
<BaseButton
  baseClassName="ms-Button"
  className="y-button y-button__color-secondary y-button__size-small"
  classNames={Object {}}
  disabled={false}
  split={false}
  styles={Object {}}
  type="button"
>
  <Block
    className=""
    textSize="small"
  >
    <span
      className="y-button--icon-wrapper-left"
    >
      <Accounts />
    </span>
    Click Me
  </Block>
</BaseButton>
`;

exports[`<Button /> with submit type matches its snapshot 1`] = `
<BaseButton
  baseClassName="ms-Button"
  className="y-button y-button__color-secondary y-button__size-regular"
  classNames={Object {}}
  disabled={false}
  split={false}
  styles={Object {}}
  type="submit"
>
  <Block
    className=""
    textSize="mediumSub"
  >
    Click Me
  </Block>
</BaseButton>
`;

exports[`<Button /> with valid href matches its snapshot 1`] = `
<BaseButton
  baseClassName="ms-Button"
  className="y-button y-button__color-secondary y-button__size-regular"
  classNames={Object {}}
  disabled={false}
  href="https://www.yammer.com"
  split={false}
  styles={Object {}}
  type=""
>
  <Block
    className=""
    textSize="mediumSub"
  >
    Click Me
  </Block>
</BaseButton>
`;
